### PR TITLE
feat: add sample_rate to spans and transactions

### PR DESCRIPTION
**Pull request by astorm**
Closes: #2023

This PR finishes [the incomplete tracing-sampling](https://github.com/elastic/apm/blob/master/specs/agents/tracing-sampling.md) spec by ensuring the sample rate value stored in the tracestate is also set on the final serialized spans and transactions. 

Preexisting Functionality: 

- A "sampled" transaction (also called "recorded") is one where spans have been generated
- A root level transaction is the first transaction in a distributed trace
- The agent makes a sampling decision for root level transactions, and the resulting boolean is stored in the `traceparent` header
- Non-root transactions use the `traceparent` header to make their sampling decision
- The sampling rate the agent used to determine if a root transaction was sampled or not is stored on the `tracestate` header
- The sample rate from `tracestate` is never used to make a sampling decision. We're only propagating it so its value can be sent to APM Server and used in metrics calculations

New Functionality:

- We've added a `sample_rate` property to transactions and spans that duplicates the value in the `tracestate` header
- There are certain scenarios, outlined [in the spec](https://github.com/elastic/apm/blob/master/specs/agents/tracing-sampling.md), where `sample_rate` is not set, or `sample_rate` is set to zero

### CODE REVIEW

File: lib/instrumentation/generic-span.js [hunk 1]
```diff
@@ -67,6 +67,17 @@ Object.defineProperty(GenericSpan.prototype, 'sampled', {
   }
 })
 
+Object.defineProperty(GenericSpan.prototype, 'sample_rate', {
+  enumerable: true,
+  get () {
+    const rate = parseFloat(this._context.tracestate.getValue('s'))
```
Comment (astorm) [inline lib/instrumentation/generic-span.js:73]: If it's a legitimate number, return the number.  Otherwise return 0.

File: lib/instrumentation/transaction.js [hunk 2]
```diff
@@ -161,6 +161,23 @@ Transaction.prototype.toJSON = function () {
     }
   }
 
+  // add sample_rate to transaction
```
Comment (astorm) [inline lib/instrumentation/transaction.js:164]: This is the edge case handling for places where we're not just passing on the sample rate. (unsampled MUST have a sample_rate of 0, invalid/empty trace state, etc.).  Tests for these edge cases start here: https://github.com/elastic/apm-agent-nodejs/pull/2038/files#diff-800eab1f0e0cb3be24e7db4feea6cd572dc4ccbce5caf260923b1828ea3ff290R47

File: test/tracecontext/sample_rate.js [hunk 3]
```diff
@@ -0,0 +1,218 @@
+const agent = require('../..').start({
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:1]: 'use strict' at the top?

File: test/tracecontext/sample_rate.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+// @return {Transaction}
+function startUnSampledTransaction () {
+  let transaction = { sampled: true }
+  let guardCount = 0
+  while (transaction.sampled && guardCount < 10000) {
+    transaction = agent.startTransaction('foo', 'bar', 'baz', 'bing')
+    guardCount++
+  }
+  if (guardCount >= 10000) {
+    throw new Error('startUnSampledTransaction could not start a sampled transaction')
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:215]:
```suggestion
    throw new Error('startUnSampledTransaction could not start an unsampled transaction')
```

File: test/tracecontext/sample_rate.js [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+const suite = require('tape')
+
+suite('Sample Rate Propagation', function (test) {
+  const TRACEPARENT_RECORDED = '00-2b9e333caa56c38e6e4d682f3184da23-a6bea2be0b62fb54-01'
+  const TRACEPARENT_NOTRECORDED = '00-c054b05d8262aaca165ed62e32898f55-4fcce7bf3ec36951-00'
+
+  test.test('sample rate is set', function (t) {
+    agent._conf.transactionSampleRate = 0.499
+    // const transaction = agent.startTransaction('foo', 'bar', 'baz', 'bing', {
+    // })
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:19]: nit: Debugging code to remove?

File: test/tracecontext/sample_rate.js [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    })
+    const span = transaction.startSpan('foo')
+    t.equals(transaction.sample_rate, 0, 'sample rate set')
+    t.ok(!span, 'span not started')
+    t.end()
+  })
+
+  // test that a sampled/recorded root transaction produces samples and
+  // transactions with a sample_rate equal to the sample rate
+  test.test('recorded root transactions have a sample rate of 0', function (t) {
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:63]: copypasta: Update this test name to match what is tested.

File: test/tracecontext/sample_rate.js [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    const transaction = agent.startTransaction('foo', 'bar', 'baz', 'bing')
+    const span = transaction.startSpan('foo')
+    t.equals(transaction.sample_rate, 0, 'sample rate set')
+    t.ok(!span, 'no span for unsampled transactions')
+    t.end()
+  })
+
+  // test that an unsampled transaction from a non-zero sample rate
+  // still produces a final serialized span with a sample rate of 0
+  test.test('recorded root transactions have a sample rate of 0', function (t) {
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:85]: copypasta: Update this test name to match what is tested.

File: test/tracecontext/sample_rate.js [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    const span = transaction.startSpan('foo')
+    const serialized = transaction.toJSON()
+    t.equals(serialized.sample_rate, 0, 'serialized sample rate is zero')
+    t.ok(!span, 'no span for unsampled transactions')
+    t.end()
+  })
+
+  // Test that an invalid tracestate in a recorded transaction
+  // results in a serialized span without a sample_rate, per the spec.
+  test.test('blank tracestate, recorded', function (t) {
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:100]: copypasta: Update this test name to match what is tested.

File: test/tracecontext/sample_rate.js [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+      t.equals(transactionSerialized.sample_rate, undefined, 'serialized transaction should have no sample_rate')
+      t.equals(spanSerialized.sample_rate, undefined, 'serialized span should have no sample_rate')
+
+      t.end()
+    })
+  })
+
+  // Test that an invalid tracestate in an unrecorded transaction
+  // results in a serialized span without a sample_rate, per the spec.
+  test.test('blank tracestate, recorded', function (t) {
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:144]: copypasta: Update this test name to match what is tested.

File: test/tracecontext/sample_rate.js [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    t.equals(serialized.sample_rate, 0, 'serialized sample rate is zero')
+    t.ok(!span, 'no span for unsampled transactions')
+    t.end()
+  })
+
+  // Test that an invalid tracestate in a recorded transaction
+  // results in a serialized span without a sample_rate, per the spec.
+  test.test('blank tracestate, recorded', function (t) {
+    // sort of gross while shenanagins to let us get an unsampled
+    // transaction with a non-zero sample rate
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:102]: Is this comment copypasta as well? It says "unsampled transaction", but this is a sampled one.

File: test/tracecontext/sample_rate.js [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+      t.end()
+    })
+  })
+
+  // Test that a blank tracestate in a recorded transaction
+  // results in a serialized span without a sample_rate, per the spec.
+  test.test('blank tracestate, recorded', function (t) {
+    // sort of gross while shenanagins to let us get an unsampled
+    // transaction with a non-zero sample rate
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:124]: copypasta?

File: test/tracecontext/sample_rate.js [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+      t.end()
+    })
+  })
+
+  // Test that an invalid tracestate in an unrecorded transaction
+  // results in a serialized span without a sample_rate, per the spec.
+  test.test('blank tracestate, recorded', function (t) {
+    // sort of gross while shenanagins to let us get an unsampled
+    // transaction with a non-zero sample rate
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:146]: Copypasta?

File: test/tracecontext/sample_rate.js [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    t.equals(transactionSerialized.sample_rate, undefined, 'serialized transaction should have no sample_rate')
+    t.ok(!span, 'no span for unsampled transaction')
+    t.end()
+  })
+
+  // Test that a blank tracestate in an unrecorded transaction
+  // results in a serialized span without a sample_rate, per the spec.
+  test.test('blank tracestate, recorded', function (t) {
+    // sort of gross while shenanagins to let us get an unsampled
+    // transaction with a non-zero sample rate
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:162]: Copypasta?

File: test/tracecontext/sample_rate.js [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+    const span = transaction.startSpan('foo')
+    const transactionSerialized = transaction.toJSON()
+    t.equals(transactionSerialized.sample_rate, undefined, 'serialized transaction should have no sample_rate')
+    t.ok(!span, 'no span for unsampled transaction')
+    t.end()
+  })
+
+  // Test that a blank tracestate in an unrecorded transaction
+  // results in a serialized span without a sample_rate, per the spec.
+  test.test('blank tracestate, recorded', function (t) {
```
Comment (trentm) [inline test/tracecontext/sample_rate.js:160]: copypasta: Update this test name to match what is tested.

File: test/tracecontext/sample_rate.js (hunk 5 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js:19]: ✅

File: test/tracecontext/sample_rate.js (hunk 3 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js:1]: ✅

File: test/tracecontext/sample_rate.js (hunk 4 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js:215]: ✅

File: test/tracecontext/sample_rate.js (hunk 6 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js:63]: ✅

File: test/tracecontext/sample_rate.js (hunk 7 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js:85]: ✅

File: lib/instrumentation/span.js [hunk 15]
```diff
@@ -219,6 +219,11 @@ Span.prototype._encode = function (cb) {
       outcome: self.outcome
     }
 
+    // if a valid sample rate is set (truthy or zero), set the property
+    if (self.sample_rate || self.sample_rate === 0) {
+      payload.sample_rate = self.sample_rate
+    }
```
[the remark this answers is not in the document]
Comment (astorm) [reply lib/instrumentation/span.js:226]: replaced with a `getTraceContextSampleRate` method, pending discussion on Tuesday.

File: lib/instrumentation/transaction.js [hunk 16]
```diff
@@ -161,6 +161,18 @@ Transaction.prototype.toJSON = function () {
...
+  // https://github.com/elastic/apm/blob/master/specs/agents/tracing-sampling.md
+  // Only set sample_rate on transaction payload if a valid trace state
+  // variable is set.
+  //
+  // "If there is no tracestate or no valid es entry with an s attribute,
+  //  then the agent must omit sample_rate from non-root transactions and
+  //  their spans."
+  if (this.sample_rate || this.sample_rate === 0) {
+    payload.sample_rate = parseFloat(this.sample_rate)
+  }
```
[the remark this answers is not in the document]
Comment (astorm) [reply lib/instrumentation/transaction.js:175]: replaced with a `getTraceContextSampleRate` method, pending discussion on Tuesday.  Also, re: the `parseFloat` -- good catch.  That was left over from a more complex version of this.

File: test/tracecontext/sample_rate.js (hunk 8 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js]: 🤦 when the comment is right but the test name is 🍝 , fixed

File: test/tracecontext/sample_rate.js (hunk 9 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js]: 🤦 when the comment is right but the test name is 🍝 , fixed

File: test/tracecontext/sample_rate.js (hunk 10 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js:101]: it was plain pasta, but a stale comment from a more complicated version.  Removing.

File: test/tracecontext/sample_rate.js (hunk 12 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js:145]: 🤦 when the comment is right but the test name is 🍝 , fixed

File: test/tracecontext/sample_rate.js (hunk 14 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js]: 🤦 when the comment is right but the test name is 🍝 , fixed

File: test/tracecontext/sample_rate.js (hunk 13 above)
Comment (astorm) [reply test/tracecontext/sample_rate.js:161]: decrufted

**Comment by astorm** [comment]
@trent per offline discussion, we're avoiding calling the property/getter multiple times in the same function and have renamed it to `sampleRate`, and have de-carbed all that test name copy/pasta. Test are passing, but back up for review.

File: test/tracecontext/sample_rate.js (hunk 13 above)
Comment (trentm) [reply test/tracecontext/sample_rate.js:157]: I meant this (and a few other) "sort of gross while shenanagins  ..." comment. That feels like it was a copypasta comment from before you factored out `start(Un)SampledTransaction()`

File: lib/instrumentation/generic-span.js [hunk 17]
```diff
@@ -67,6 +67,17 @@ Object.defineProperty(GenericSpan.prototype, 'sampled', {
   }
 })
 
+Object.defineProperty(GenericSpan.prototype, 'sampleRate', {
+  enumerable: true,
+  get () {
+    const rate = parseFloat(this._context.tracestate.getValue('s'))
+    if (rate || rate === 0) {
```
Comment (trentm) [inline lib/instrumentation/generic-span.js:74]: 1. Should this discard values outside of `[0, 1]` as invalid?  The *comments* in apm-server (https://github.com/elastic/apm-server/blob/b92191369f58c939ac894b478f1bb6385c07dd63/model/modeldecoder/v2/model.go#L768-L771) and the spec suggest it should be in that range, though it doesn't look to me that apm-server validates that.

2. nit: Also perhaps use `if (!isNaN(rate)) {`  to guard the result from `parseFloat()`.

File: lib/instrumentation/span.js [hunk 18]
```diff
@@ -219,6 +219,12 @@ Span.prototype._encode = function (cb) {
       outcome: self.outcome
     }
 
+    // if a valid sample rate is set (truthy or zero), set the property
+    const sampleRate = self.sampleRate
+    if (sampleRate || sampleRate === 0) {
```
Comment (trentm) [inline lib/instrumentation/span.js:224]: nit: Perhaps use `if (sampleRate !== null)` given that the sampleRate getting returns `null` for an invalid value.

File: lib/instrumentation/transaction.js [hunk 19]
```diff
@@ -161,6 +161,19 @@ Transaction.prototype.toJSON = function () {
...
+  // add sample_rate to transaction
+  // https://github.com/elastic/apm/blob/master/specs/agents/tracing-sampling.md
+  // Only set sample_rate on transaction payload if a valid trace state
+  // variable is set.
+  //
+  // "If there is no tracestate or no valid es entry with an s attribute,
+  //  then the agent must omit sample_rate from non-root transactions and
+  //  their spans."
+  const sampleRate = this.sampleRate
+  if (sampleRate || sampleRate === 0) {
```
Comment (trentm) [inline lib/instrumentation/transaction.js:173]: nit: Perhaps use `if (sampleRate !== null)` given that the sampleRate getting returns `null` for an invalid value.

File: lib/instrumentation/generic-span.js (hunk 17 above)
Comment (trentm) [reply lib/instrumentation/generic-span.js:74]: FWIW, I'm fine punting the range check to a separate issue .

Comment (astorm) [reply lib/instrumentation/generic-span.js:74]: Longer term it's probably worth thinking about where/when values that could be validated should be validated. 

For now, validating this value here seems fine, as it also ends up protecting against `NaN`.

```
    if(rate >=0 || rate <= 1) {
      return rate
    }
```

File: lib/instrumentation/transaction.js (hunk 19 above)
Comment (astorm) [reply lib/instrumentation/transaction.js:173]: ✅

File: lib/instrumentation/generic-span.js [hunk 20]
```diff
@@ -71,7 +71,7 @@ Object.defineProperty(GenericSpan.prototype, 'sampleRate', {
   enumerable: true,
   get () {
     const rate = parseFloat(this._context.tracestate.getValue('s'))
-    if (rate || rate === 0) {
+    if (rate >= 0 || rate <= 1) {
```
Comment (trentm) [inline lib/instrumentation/generic-span.js:74]:
```suggestion
    if (rate >= 0 && rate <= 1) {
```

You want `&&` here, I think.

File: lib/instrumentation/span.js (hunk 18 above)
Comment (trentm) [inline lib/instrumentation/span.js:224]:
```suggestion
    if (sampleRate !== null) {
```

Do the !== null check here as well?

**Review by trentm — CHANGES_REQUESTED** [review]

**Review by trentm — APPROVED** [review]
